doc(go-client): add per config detail config reference page (#6004)
ClickHouse/clickhouse-docs: docs/integrations/language-clients/go/config-reference.md

---
sidebar_label: 'Configuration reference'
sidebar_position: 3
keywords: ['clickhouse', 'go', 'golang', 'configuration', 'options', 'reference', 'DSN', 'connection pool', 'TLS', 'compression', 'timeout']
description: 'Complete per-option configuration reference for the clickhouse-go client, covering connection-level, context-level, and batch options.'
slug: /integrations/language-clients/go/config-reference
title: 'Configuration reference'
doc_type: 'reference'
---

# Configuration reference {#config-reference}

This page documents every configurable option in `clickhouse-go` v2.x. For narrative guides with code examples, see [Configuration](/integrations/language-clients/go/configuration).

## How options are set {#how-options-are-set}

Options exist at three scopes:

| Scope | How to set | Lifetime |
|-------|-----------|----------|
| **Connection** | `clickhouse.Options` struct or DSN string | All queries on the connection |
| **Query** | `clickhouse.Context()` with `WithXxx` functions | Single query execution |

> | **Batch** | `PrepareBatch()` option functions | Single batch operation |

Reviewer: Should we call out which one takes precedence?

    DialTimeout: 10 * time.Second,
    Compression: &clickhouse.Compression{Method: clickhouse.CompressionLZ4},
})
```

**Via DSN string:**

```go
db, err := sql.Open("clickhouse", "clickhouse://user:pass@localhost:9000/default?dial_timeout=10s&compress=lz4")
```

**Via Connector (database/sql with Options struct):**

```go
db := sql.OpenDB(clickhouse.Connector(&clickhouse.Options{
    Addr:        []string{"localhost:9000"},
    Auth:        clickhouse.Auth{Database: "default", Username: "default"},
    DialTimeout: 10 * time.Second,
}))
// Set database/sql-only pool settings after creation
db.SetConnMaxIdleTime(5 * time.Minute)
```

**Via context (per-query):**

```go
ctx := clickhouse.Context(context.Background(),
    clickhouse.WithQueryID("my-query-123"),
    clickhouse.WithSettings(clickhouse.Settings{"max_execution_time": 60}),
)
rows, err := conn.Query(ctx, "SELECT ...")
```

---

## Connection options {#connection-options}

### Protocol and connection {#protocol-and-connection}

| Option | Type | Default | DSN param | Description | Best practice | When misconfigured |
|--------|------|---------|-----------|-------------|---------------|-------------------|

> | `Protocol` | `Protocol` (int) | `Native` | Scheme: `clickhouse://`=Native, `http://`=HTTP | Communication protocol: `Native` (0) for TCP, `HTTP` (1) for HTTP | Use Native for ~30% better performance. Use HTTP for proxy support, firewall traversal (port 80/443), or HTTP-only compression (`gzip`/`br`). See [TCP vs HTTP](/integrations/language-clients/go/configuration#tcp-vs-http). | HTTP scheme with Native port (9000): connection refused. Native blocked by firewall: timeouts. |

Reviewer: The current structure works well for AI agents. @Blargian 

<img width="2520" height="654" alt="Image" src="https://github.com/user-attachments/assets/42382c2d-4805-42a6-aefc-cb26e227ebbd" />

I’d like to know your recommendations for improving readability for both human and AI agents. 
Probably, we should follow the same pattern that Core uses ? https://clickhouse.com/docs/operations/server-configuration-parameters/settings

Reviewer 2: From what I've read agents process markdown tables just as well. I find them more readable for humans too, but could just be a personal preference.

Author: I would leave it as markdown tables for now :) thanks

### Authentication {#authentication}

| Option | Type | Default | DSN param | Description | Best practice | When misconfigured |
|--------|------|---------|-----------|-------------|---------------|-------------------|
| `Auth.Username` | `string` | `"default"` | `username` or URL user portion | Username for ClickHouse authentication | Never use `default` in production. Create dedicated users with minimal permissions. | Wrong username: `"Code: 516. DB::Exception: Authentication failed"`. Empty string: silently uses `"default"`. |
| `Auth.Password` | `string` | `""` | `password` or URL password portion | Password for ClickHouse authentication | Use env vars or secret managers in production. URL-encode special characters in DSN. | Wrong password: `"Code: 516. DB::Exception: Authentication failed"`. Special chars not URL-encoded: parsing errors. |
| `Auth.Database` | `string` | `""` (server default) | `database` or URL path (`/mydb`) | Default database for the connection | Always specify explicitly. Use dedicated databases per application in production. | Non-existent: `"Code: 81. DB::Exception: Database xyz doesn't exist"`. Empty in multi-tenant setup: queries hit wrong database. |

> | `GetJWT` | `func(ctx) (string, error)` | `nil` | — (programmatic only) | Callback returning JWT for ClickHouse Cloud auth. Overridable per query with `WithJWT(token)`. | Implement token caching/refresh -- called per connection/request. | Expired token: auth errors. Blocking callback: timeouts. JWT takes precedence over user/pass. Requires TLS -- without it, falls back to user/pass silently. |

Reviewer: For the recently introduced settings, it would be beneficial to include a client version when it was first added

Author: Done. Added version for configs and settings since v2.35

```go
GetJWT: func(ctx context.Context) (string, error) {
    return getTokenFromVault(ctx)
}
```

---

### Timeouts {#timeouts}

> | Option | Type | Default | DSN param | Description | Best practice | When misconfigured |

Reviewer: ❤️ to this `when misconfigured` section

|--------|------|---------|-----------|-------------|---------------|-------------------|
| `DialTimeout` | `time.Duration` | `30s` | `dial_timeout` | Max time to establish a new connection. Also controls pool acquisition wait when `MaxOpenConns` is reached. | 5-10s on LAN, 15-30s on WAN/cloud. Never below 1s. | Too short: `"clickhouse: acquire conn timeout"` during congestion. Too long (> 60s): app hangs during outages. |

> | `ReadTimeout` | `time.Duration` | `5m` (300s) | `read_timeout` | Max time to wait for a server response per read call. Applied per block, not entire query. Context deadline takes precedence. | 10-30s for OLTP, 5-30m for OLAP/long queries. | Too short: `"i/o timeout"` or `"read: connection reset by peer"` mid-query; server continues executing. Too long: dead connections not detected. |

Reviewer: what is `for OLTP` here?

Author: Updated the wording. I was meaning to say short queries vs long-running queries.

    Logger: logger,
    // ...
})
```

See [Logging](/integrations/language-clients/go/configuration#logging) for full examples.

---

### Buffers and memory {#buffers-and-memory}

| Option | Type | Default | DSN param | Per-query | Description | Best practice | When misconfigured |
|--------|------|---------|-----------|-----------|-------------|---------------|-------------------|
| `BlockBufferSize` | `uint8` | `2` | `block_buffer_size` | Yes (`WithBlockBufferSize`) | Decoded blocks to buffer when reading results. Enables concurrent read + decode. | Default 2 is fine. 5-10 for large streaming results. Memory = buffer x block size x concurrent queries. | Too small (1): blocks reader, higher latency. Too large (> 50): high memory, diminishing returns. |
| `FreeBufOnConnRelease` | `bool` | `false` | — | No | Release connection memory buffer after each query instead of reusing. | `false` for high query rates. `true` in memory-constrained containers or infrequent large batches. | `false` + limited memory: buffers accumulate (memory = buffer x idle conns). `true` + high rate: GC pressure, increased CPU. |

---

### HTTP-specific {#http-specific}

> These options only affect `Protocol: clickhouse.HTTP`. Silently ignored for Native.

Reviewer: Should it be wrapped as a `warning`?

Author: Yes. updated.

| `CustomSetting` | `CustomSetting{Value string}` | — | — | Yes (via `WithSettings`) | Marks a setting as "custom" (non-important) for Native protocol. Won't error if server doesn't recognize it. HTTP treats all settings as custom by default. | Use for experimental or version-specific settings. | Marking important settings as custom: silently ignored if unsupported. |

**Common settings:**

| Setting | Type | Description |
|---------|------|-------------|
| `max_execution_time` | int | Query timeout in seconds |
| `max_memory_usage` | int | Memory limit per query (bytes) |
| `max_block_size` | int | Block size for processing |
| `readonly` | int | 1 = read-only, 2 = read-only + settings changes |

```go
Settings: clickhouse.Settings{
    "max_execution_time":  60,                                        // important -- errors if unknown

>     "my_custom_setting":   clickhouse.CustomSetting{Value: "value"},  // custom -- ignored if unknown

Reviewer: 👍

        log.Printf("Progress: %d rows, %d bytes", p.Rows, p.Bytes)
    }),
)
rows, err := conn.Query(ctx, "SELECT * FROM users WHERE id = {user_id:String}")
```

---

## Batch options {#batch-options}

Passed to `PrepareBatch()`. Import: `github.com/ClickHouse/clickhouse-go/v2/lib/driver`.

| Option | Default | Description | Best practice | When misconfigured |
|--------|---------|-------------|---------------|-------------------|
| `WithReleaseConnection` | Connection held until `Send()` | Release connection to pool immediately after `PrepareBatch()`. Re-acquires on `Send()`/`Flush()`. | Use for long-lived batches (minutes/hours) to prevent pool exhaustion. | Not using for long batches: `"acquire conn timeout"` if many active. |
| `WithCloseOnFlush` | Batch stays open | Auto-close batch when `Flush()` is called. | Use for one-shot batches. Saves explicit `Close()`. | Using with multiple `Flush()` calls: first flush closes batch, subsequent ops fail. |

```go
batch, err := conn.PrepareBatch(ctx, "INSERT INTO table",
    driver.WithReleaseConnection(),
    driver.WithCloseOnFlush(),
)
```

---

## Quick reference tables {#quick-reference-tables}

> ### Connection pool sizing {#pool-sizing}

Reviewer: ```suggestion
### Connection pool sizing recommendations {#pool-sizing}
```

| `max_idle_conns` | `MaxIdleConns` | `?max_idle_conns=20` |
| `conn_max_lifetime` | `ConnMaxLifetime` | `?conn_max_lifetime=30m` |
| `connection_open_strategy` | `ConnOpenStrategy` | `?connection_open_strategy=round_robin` |
| `block_buffer_size` | `BlockBufferSize` | `?block_buffer_size=10` |
| `compress` | `Compression.Method` | `?compress=lz4` |
| `compress_level` | `Compression.Level` | `?compress_level=6` |
| `max_compression_buffer` | `MaxCompressionBuffer` | `?max_compression_buffer=20971520` |
| `secure` | `TLS` | `?secure=true` |
| `skip_verify` | `TLS.InsecureSkipVerify` | `?skip_verify=true` |
| `debug` | `Debug` | `?debug=true` |
| `client_info_product` | `ClientInfo.Products` | `?client_info_product=myapp/1.0` |
| `http_proxy` | `HTTPProxyURL` | `?http_proxy=http%3A%2F%2Fproxy%3A8080` |
| `http_path` | `HttpUrlPath` | `?http_path=/clickhouse` |
| *(any other)* | `Settings[key]` | `?max_execution_time=60` |

---

## Troubleshooting {#troubleshooting}

### "acquire conn timeout" {#acquire-conn-timeout}

**Cause:** Connection pool exhausted -- all `MaxOpenConns` connections are in use and none became available within `DialTimeout`.

> **Fix:**

Reviewer: Should one take each step one-by-one or do it all at once?

---
sidebar_label: 'Configuration reference'
sidebar_position: 3
keywords: ['clickhouse', 'go', 'golang', 'configuration', 'options', 'reference', 'DSN', 'connection pool', 'TLS', 'compression', 'timeout']
description: 'Complete per-option configuration reference for the clickhouse-go client, covering connection-level, context-level, and batch options.'
slug: /integrations/language-clients/go/config-reference

> title: 'Configuration reference'

Reviewer 2: ```suggestion
title: 'Go-client configuration reference'
```

Search indexes most heavily on title and H{1-4} elements so it's good to include which client it is configuration reference for,  otherwise it could surface as a top result for a search containing "configuration" or "reference"

doc_type: 'reference'
---

> # Configuration reference {#config-reference}

Reviewer 2: ```suggestion
```

Docusaurus renders the H1 element from the front-matter `title` so not needed here

This page documents every configurable option in `clickhouse-go` v2.x. For narrative guides with code examples, see [Configuration](/integrations/language-clients/go/configuration).

:::note Version annotations
Options added in `clickhouse-go` v2.35.0 or later are marked with *(Since vX.Y.Z)* next to their description. Options without a "Since" tag have been available since v2.0 and are present in every supported release.

> :::

Reviewer 2: ```suggestion
This page documents every configurable option in `clickhouse-go` v2.x. For a guide with code examples, see [Configuration](/integrations/language-clients/go/configuration).

:::note[Version annotations]
Options added in `clickhouse-go` v2.35.0 or later are marked with *(Since vX.Y.Z)* next to their description. Options without a "Since" tag have been available since v2.0 and are present in every supported release.
:::
```

### Authentication {#authentication}

| Option | Type | Default | DSN param | Description | Best practice | When misconfigured |
|--------|------|---------|-----------|-------------|---------------|-------------------|
| `Auth.Username` | `string` | `"default"` | `username` or URL user portion | Username for ClickHouse authentication | Never use `default` in production. Create dedicated users with minimal permissions. | Wrong username: `"Code: 516. DB::Exception: Authentication failed"`. Empty string: silently uses `"default"`. |
| `Auth.Password` | `string` | `""` | `password` or URL password portion | Password for ClickHouse authentication | Use env vars or secret managers in production. URL-encode special characters in DSN. | Wrong password: `"Code: 516. DB::Exception: Authentication failed"`. Special chars not URL-encoded: parsing errors. |
| `Auth.Database` | `string` | `""` (server default) | `database` or URL path (`/mydb`) | Default database for the connection | Always specify explicitly. Use dedicated databases per application in production. | Non-existent: `"Code: 81. DB::Exception: Database xyz doesn't exist"`. Empty in multi-tenant setup: queries hit wrong database. |

> | `GetJWT` | `func(ctx) (string, error)` | `nil` | — (programmatic only) | Callback returning JWT for ClickHouse Cloud auth. Overridable per query with `WithJWT(token)`. *(Since v2.35.0)* | Implement token caching/refresh -- called per connection/request. | Expired token: auth errors. Blocking callback: timeouts. JWT takes precedence over user/pass. Requires TLS -- without it, falls back to user/pass silently. |

Reviewer 2: ```suggestion
| `GetJWT` | `func(ctx) (string, error)` | `nil` | (programmatic only) | Callback returning JWT for ClickHouse Cloud auth. Overridable per query with `WithJWT(token)`. *(Since v2.35.0)* | Implement token caching/refresh - called per connection/request. | Expired token: auth errors. Blocking callback: timeouts. JWT takes precedence over user/pass. Requires TLS - without it, falls back to user/pass silently. |
```

| `ConnMaxLifetime` | `time.Duration` | `1h` | `conn_max_lifetime` | Both | Max duration a connection can be reused. Checked on return to pool. | 1-5h stable envs. 5-15m for K8s/rolling deploys. Never infinite. | Too short (&#60; 1m): churn, higher latency. Too long/infinite: stale connections, DNS changes not picked up, traffic never rebalances. |
| `ConnMaxIdleTime` | `time.Duration` | `0` (none) | — | `database/sql` only | Max time a connection can sit *idle* before closing. Not in `Options` struct -- set via `db.SetConnMaxIdleTime()`. | 5-10m for K8s/bursty workloads to reclaim idle connections after traffic spikes. | Not set: idle connections persist until `ConnMaxLifetime`. Too short (&#60; 30s): connections recreated during normal gaps. |

> :::note database/sql only

Reviewer 2: ```suggestion
:::note[Database/SQL only]
```

Author: This is the Go package name. has to be in lower case unfortunately.

> `ConnMaxIdleTime` is a standard Go `database/sql` pool setting. It is not available in the `clickhouse.Options` struct or via `clickhouse.Open()`. Set it after `OpenDB()`:

Reviewer 2: ```suggestion
`ConnMaxIdleTime` is a standard Go `database/sql` pool setting. It isn't available in the `clickhouse.Options` struct or via `clickhouse.Open()`. Set it after `OpenDB()`:
```

```go
db := clickhouse.OpenDB(&clickhouse.Options{...})
db.SetConnMaxIdleTime(5 * time.Minute)
```
:::

See [Connection Pooling](/integrations/language-clients/go/configuration#connection-pooling) for usage details.

---

### Standard database/sql pool settings {#sql-db-settings}

When using `clickhouse.OpenDB()` or `sql.Open("clickhouse", dsn)`, the returned `*sql.DB` supports Go's standard pool methods. `OpenDB()` auto-applies the first three from `Options`:

| Method | Options equivalent | Notes |
|--------|--------------------|-------|
| `db.SetMaxIdleConns(n)` | `MaxIdleConns` | Auto-applied by `OpenDB()` |
| `db.SetMaxOpenConns(n)` | `MaxOpenConns` | Auto-applied by `OpenDB()` |
| `db.SetConnMaxLifetime(d)` | `ConnMaxLifetime` | Auto-applied by `OpenDB()` |
| `db.SetConnMaxIdleTime(d)` | *None* | Must be set manually post-creation |

> :::note ClickHouse API (clickhouse.Open)

Reviewer 2: ```suggestion
:::note[ClickHouse API (clickhouse.Open)]
```

> These methods are NOT available on the connection returned by `clickhouse.Open()`. The ClickHouse API manages its own pool internally using the `Options` struct fields directly.

Reviewer 2: ```suggestion
These methods are **not** available on the connection returned by `clickhouse.Open()`. The ClickHouse API manages its own pool internally using the `Options` struct fields directly.
```

| Option | Type | Default | DSN param | Per-query | Description | Best practice | When misconfigured |
|--------|------|---------|-----------|-----------|-------------|---------------|-------------------|
| `BlockBufferSize` | `uint8` | `2` | `block_buffer_size` | Yes (`WithBlockBufferSize`) | Decoded blocks to buffer when reading results. Enables concurrent read + decode. | Default 2 is fine. 5-10 for large streaming results. Memory = buffer x block size x concurrent queries. | Too small (1): blocks reader, higher latency. Too large (> 50): high memory, diminishing returns. |
| `FreeBufOnConnRelease` | `bool` | `false` | — | No | Release connection memory buffer after each query instead of reusing. | `false` for high query rates. `true` in memory-constrained containers or infrequent large batches. | `false` + limited memory: buffers accumulate (memory = buffer x idle conns). `true` + high rate: GC pressure, increased CPU. |

---

### HTTP-specific {#http-specific}

:::warning Silently ignored on Native
These options only affect `Protocol: clickhouse.HTTP`. They are silently ignored when using the Native protocol — no error or warning is emitted.

> :::

Reviewer 2: ```suggestion
:::warning[Silently ignored on Native]
These options only affect `Protocol: clickhouse.HTTP`. They're silently ignored when using the Native protocol, and no error or warning is emitted.
:::
```

| `HTTPProxyURL` | `*url.URL` | `nil` (uses env vars) | `http_proxy` (URL-encoded) | HTTP proxy for routing requests | Set explicitly if proxy required. Overrides `HTTP_PROXY`/`HTTPS_PROXY` env vars. | Wrong address: `"dial tcp: lookup proxy: no such host"`. Proxy needs auth: HTTP 407. |
| `TransportFunc` | `func(*http.Transport) (http.RoundTripper, error)` | `nil` | — | Custom HTTP transport factory. Receives default transport for wrapping. *(Since v2.41.0)* | Use for observability middleware. Don't override `Proxy`, `DialContext`, `TLSClientConfig`. | Returning `nil`: panic. Overriding client fields: TLS/proxy silently ignored. Blocking RoundTripper: deadlocks. |

:::note Two-layer HTTP pooling
When using HTTP, there are two connection pools:
- **Layer 1 (application):** `MaxIdleConns` / `MaxOpenConns` -- controls `httpConnect` objects
- **Layer 2 (transport):** `HttpMaxConnsPerHost` -- controls underlying TCP connections

The Native protocol has a simple 1:1 mapping and ignores `HttpMaxConnsPerHost`.

> :::

Reviewer 2: ```suggestion
:::note[Two-layer HTTP pooling]
When using HTTP, there are two connection pools:
- **Layer 1 (application):** `MaxIdleConns` / `MaxOpenConns` -- controls `httpConnect` objects
- **Layer 2 (transport):** `HttpMaxConnsPerHost` -- controls underlying TCP connections

The Native protocol has a simple 1:1 mapping and ignores `HttpMaxConnsPerHost`.
:::
```

**Common settings:**

| Setting | Type | Description |
|---------|------|-------------|
| `max_execution_time` | int | Query timeout in seconds |
| `max_memory_usage` | int | Memory limit per query (bytes) |
| `max_block_size` | int | Block size for processing |
| `readonly` | int | 1 = read-only, 2 = read-only + settings changes |

```go
Settings: clickhouse.Settings{
    "max_execution_time":  60,                                        // important -- errors if unknown
    "my_custom_setting":   clickhouse.CustomSetting{Value: "value"},  // custom -- ignored if unknown
}
```

---

## Context-level query options {#context-options}

Set per-query using `clickhouse.Context()`:

```go
ctx := clickhouse.Context(context.Background(),
    clickhouse.WithQueryID("my-query"),
    clickhouse.WithSettings(clickhouse.Settings{"max_execution_time": 60}),
)
```

:::note Context deadline behavior
If the context has a deadline > 1s, `max_execution_time` is automatically set to `seconds_remaining + 5`. This overrides any manually set value.

> :::

Reviewer 2: ```suggestion
:::note[Context deadline behavior]
If the context has a deadline > 1s, `max_execution_time` is automatically set to `seconds_remaining + 5`. This overrides any manually set value.
:::
```

| Option | Type | Default | Protocol | Description | Best practice | When misconfigured |
|--------|------|---------|----------|-------------|---------------|-------------------|
| `WithQueryID` | `string` | Auto-generated | Both | Custom query identifier. Visible in `system.query_log` and `system.processes`. | Use UUIDs. Useful for `KILL QUERY WHERE query_id='...'`. | Duplicate IDs: confusion in `system.query_log`. |
| `WithQuotaKey` | `string` | `""` | Both | Quota key for multi-tenant resource limits. Requires server-side quota config. | Use for per-customer/per-user limits. | Quota not configured: silently ignored. |
| `WithJWT` | `string` | `""` | HTTPS only | Per-query JWT override for ClickHouse Cloud. *(Since v2.35.0)* | Use for per-request auth in multi-tenant proxies. | Without TLS: ignored, falls back to connection auth. Expired: `"Token has expired"`. |
| `WithSettings` | `Settings` | Inherits connection | Both | Per-query server settings. Merged with connection settings; context wins on conflict. | Override `max_execution_time` or `max_rows_to_read` per query type. | Same as connection-level `Settings`. |

> | `WithParameters` | `Parameters` (`map[string]string`) | `nil` | Both | Server-side parameterized query values. Query syntax: `{param_name:Type}`. | Use instead of string concatenation for SQL injection safety. | Missing param: `"Substitution {param_name:Type} is not set"`. Wrong type: `"Cannot parse string 'abc' as UInt64"`. |

Reviewer 2: ```suggestion
| `WithParameters` | `Parameters` (`map[string]string`) | `nil` | Both | Server-side parameterized query values. Query syntax: `{param_name:Type}`. | Use instead of string concatenation for SQL injection safety. | Missing param: `"Substitution {param_name:Type} isn't set"`. Wrong type: `"Cannot parse string 'abc' as UInt64"`. |
```

| `read_timeout` | `ReadTimeout` | `?read_timeout=5m` |
| `max_open_conns` | `MaxOpenConns` | `?max_open_conns=50` |
| `max_idle_conns` | `MaxIdleConns` | `?max_idle_conns=20` |
| `conn_max_lifetime` | `ConnMaxLifetime` | `?conn_max_lifetime=30m` |
| `connection_open_strategy` | `ConnOpenStrategy` | `?connection_open_strategy=round_robin` |
| `block_buffer_size` | `BlockBufferSize` | `?block_buffer_size=10` |
| `compress` | `Compression.Method` | `?compress=lz4` |
| `compress_level` | `Compression.Level` | `?compress_level=6` |
| `max_compression_buffer` | `MaxCompressionBuffer` | `?max_compression_buffer=20971520` |
| `secure` | `TLS` | `?secure=true` |
| `skip_verify` | `TLS.InsecureSkipVerify` | `?skip_verify=true` |
| `debug` | `Debug` | `?debug=true` |
| `client_info_product` | `ClientInfo.Products` | `?client_info_product=myapp/1.0` |
| `http_proxy` | `HTTPProxyURL` | `?http_proxy=http%3A%2F%2Fproxy%3A8080` |
| `http_path` | `HttpUrlPath` | `?http_path=/clickhouse` |
| *(any other)* | `Settings[key]` | `?max_execution_time=60` |

---

## Troubleshooting {#troubleshooting}

### Connection pool exhausted: "acquire conn timeout" {#acquire-conn-timeout}

> **Cause:** Connection pool exhausted -- all `MaxOpenConns` connections are in use and none became available within `DialTimeout`.

Reviewer 2: ```suggestion
**Cause:** Connection pool exhausted - all `MaxOpenConns` connections are in use and none became available within `DialTimeout`.
```

**Fix (try in order — diagnose root cause before tuning knobs):**
1. Check for long-running queries holding connections: `SELECT query_id, elapsed FROM system.processes ORDER BY elapsed DESC`. If found, address the slow queries first.
2. If you run long-lived batches (minutes/hours between `PrepareBatch()` and `Send()`), use `WithReleaseConnection()` to return the connection to the pool while the batch is open.
3. Increase `MaxOpenConns` to match observed concurrency.

> 4. Increase `DialTimeout` only if bursts are expected and acquisition wait is the actual bottleneck.

Reviewer 2: ```suggestion
**Fix**

Try the following steps in order, and diagnose the root cause before tuning knobs:

1. Check for long-running queries holding connections: `SELECT query_id, elapsed FROM system.processes ORDER BY elapsed DESC`. If found, address the slow queries first.
2. If you run long-lived batches (minutes/hours between `PrepareBatch()` and `Send()`), use `WithReleaseConnection()` to return the connection to the pool while the batch is open.
3. Increase `MaxOpenConns` to match observed concurrency.
4. Increase `DialTimeout` only if bursts are expected and acquisition wait is the actual bottleneck.
```
